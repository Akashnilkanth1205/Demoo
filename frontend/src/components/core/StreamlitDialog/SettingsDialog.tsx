/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ChangeEvent, PureComponent, ReactNode } from "react"
import UISelectbox from "components/shared/Dropdown"
import { CustomThemeConfig } from "autogen/proto"
import {
  createPresetThemes,
  createTheme,
  toThemeInput,
  ThemeConfig,
} from "theme"
import Modal, { ModalHeader, ModalBody } from "components/shared/Modal"
import ThemeCreator from "./ThemeCreator"
import { UserSettings } from "./UserSettings"
import { StyledHeader, StyledLabel, StyledSmall } from "./styled-components"

export interface Props {
  isServerConnected: boolean
  onClose: () => void
  onSave: (settings: UserSettings) => void
  settings: UserSettings
  allowRunOnSave: boolean
  allowedThemes: ThemeConfig[]
}

/**
 * Implements a dialog that is used to configure user settings.
 */
export class SettingsDialog extends PureComponent<Props, UserSettings> {
  private activeSettings: UserSettings

  constructor(props: Props) {
    super(props)
    // Holds the settings that will be saved when the "save" button is clicked.
    this.state = { ...this.props.settings }

    // Holds the actual settings that Streamlit is using.
    this.activeSettings = { ...this.props.settings }
  }

  public render = (): ReactNode => {
    const themeIndex = this.props.allowedThemes.findIndex(
      theme => theme.name === this.activeSettings.activeTheme.name
    )
    const hasCustomTheme =
      this.props.allowedThemes.length !== createPresetThemes().length

    return (
      <Modal isOpen onClose={this.handleCancelButtonClick}>
        <ModalHeader>Settings</ModalHeader>
        <ModalBody>
          {this.props.allowRunOnSave ? (
            <>
              <StyledHeader>Development</StyledHeader>
              <label>
                <input
                  disabled={!this.props.isServerConnected}
                  type="checkbox"
                  name="runOnSave"
                  checked={
                    this.state.runOnSave && this.props.isServerConnected
                  }
                  onChange={this.handleCheckboxChange}
                />{" "}
                Run on save
              </label>
              <br />
              <StyledSmall>
                Automatically updates the app when the underlying code is
                updated
              </StyledSmall>
            </>
          ) : null}
          <h3>Appearance</h3>
          <label>
            <input
              type="checkbox"
              name="wideMode"
              checked={this.state.wideMode}
              onChange={this.handleCheckboxChange}
            />{" "}
            Wide mode
          </label>
          <div>
            <StyledSmall>
              Turn on to make this app occupy the entire width of the screen
            </StyledSmall>
          </div>
          {this.props.allowedThemes.length > 1 ? (
            <>
              <StyledLabel>Theme</StyledLabel>
              <StyledSmall>Choose app and font colors/styles</StyledSmall>
              <UISelectbox
                options={this.props.allowedThemes.map(theme => theme.name)}
                disabled={false}
                onChange={this.handleThemeChange}
                value={themeIndex}
              />
              <ThemeCreator
                label={
                  hasCustomTheme
                    ? "Edit Existing Custom Theme"
                    : "Create a new Custom Theme"
                }
                updateThemeInput={this.handleThemeCreator}
                themeInput={{
                  ...toThemeInput(this.state.activeTheme.emotion),
                  name: hasCustomTheme
                    ? this.state.activeTheme.name
                    : "Custom theme",
                }}
              />
            </>
          ) : null}
        </ModalBody>
      </Modal>
    )
  }

  public componentDidMount(): void {
    this.setState({ ...this.activeSettings })
  }

  private changeSingleSetting = (name: string, value: boolean): void => {
    // TypeScript doesn't currently have a good solution for setState with
    // a dynamic key name:
    // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/26635
    this.setState(state => ({ ...state, [name]: value }), this.saveSettings)
  }

  private handleCheckboxChange = (e: ChangeEvent<HTMLInputElement>): void => {
    this.changeSingleSetting(e.target.name, e.target.checked)
  }

  private handleThemeChange = (index: number): void => {
    this.setState(
      {
        activeTheme: this.props.allowedThemes[index],
      },
      this.saveSettings
    )
  }

  private handleThemeCreator = (
    themeInput: Partial<CustomThemeConfig>
  ): void => {
    this.setState({ activeTheme: createTheme(themeInput) }, this.saveSettings)
  }

  private handleCancelButtonClick = (): void => {
    // Discard settings from this.state by not saving them in this.settings.
    // this.settings = {...this.state};
    this.props.onClose()
  }

  private saveSettings = (): void => {
    this.activeSettings = { ...this.state }
    this.props.onSave(this.activeSettings)
  }
}
