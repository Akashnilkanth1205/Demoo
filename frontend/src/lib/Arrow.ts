/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"

import { IStyler } from "autogen/proto"

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

interface TableDimensions {
  headerRows: number
  headerColumns: number
  dataRows: number
  dataColumns: number
  rows: number
  columns: number
}

type TableCellType = "blank" | "index" | "columns" | "data"

interface TableCell {
  type: TableCellType
  id?: string
  classNames: string
  content: string
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: Arrow
}

export class Arrow {
  private table: Table

  private schema: Schema

  private index: string[][]

  private columns: string[][]

  private data: any[][]

  private styler?: Styler

  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
    this.table = Table.from(dataBuffer)
    this.schema = this.getSchema()
    this.index = this.getIndex()
    this.columns = this.getColumns()
    this.data = this.getData()
    this.styler = {
      uuid: styler?.uuid || undefined,
      caption: styler?.caption || undefined,
      styles: styler?.styles || undefined,
      displayValues: styler?.displayValues
        ? new Arrow(styler.displayValues)
        : undefined,
    }
  }

  private getSchema(): Schema {
    const schema = this.table.schema.metadata.get("pandas")
    if (schema == null) {
      throw new Error("Table schema is missing.")
    }
    return JSON.parse(schema)
  }

  private static getColumnData(column: Column): any[] {
    return range(0, column.length).map(rowIndex => column.get(rowIndex))
  }

  private getIndex(): string[][] {
    return this.schema.index_columns.map(field => {
      const isRangeIndex = typeof field === "object" && field.kind === "range"
      if (isRangeIndex) {
        const { start, stop, step } = field as RangeIndex
        return range(start, stop, step).map(String)
      }
      const column = this.table.getColumn(field as string)
      return Arrow.getColumnData(column).map(String)
    })
  }

  private getColumns(): string[][] {
    const isMultiIndex = this.schema.column_indexes.length > 1
    return unzip(
      this.schema.columns
        .map(column => column.field_name)
        .filter(fieldName => !this.schema.index_columns.includes(fieldName))
        .map(fieldName =>
          isMultiIndex
            ? JSON.parse(
                fieldName
                  .replace(/\(/g, "[")
                  .replace(/\)/g, "]")
                  .replace(/'/g, '"')
              )
            : [fieldName]
        )
    )
  }

  private getData(): any[][] {
    const rows = this.table.length
    const columns = this.columns.length > 0 ? this.columns[0].length : 0

    return range(0, rows).map(rowIndex =>
      range(0, columns).map(columnIndex =>
        this.table
          .getColumnAt(columnIndex)
          ?.get(rowIndex)
          ?.toString()
      )
    )
  }

  get tableId(): string | undefined {
    return this.styler?.uuid && `T_${this.styler.uuid}`
  }

  get tableStyles(): string | undefined {
    return this.styler?.styles
  }

  get caption(): string | undefined {
    return this.styler?.caption
  }

  get dimensions(): TableDimensions {
    const [headerColumns, dataRowsCheck] = this.index.length
      ? [this.index.length, this.index[0].length]
      : [0, 0]

    const [headerRows, dataColumnsCheck] = this.columns.length
      ? [this.columns.length, this.columns[0].length]
      : [0, 0]

    const [dataRows, dataColumns] = this.data.length
      ? [this.data.length, this.data[0].length]
      : // If there is no data, default to the number of header columns.
        [0, dataColumnsCheck]

    if (
      (dataRows !== 0 && dataRows !== dataRowsCheck) ||
      (dataColumns !== 0 && dataColumns !== dataColumnsCheck)
    ) {
      throw new Error(
        "Table dimensions don't align: " +
          `rows(${dataRows} != ${dataRowsCheck}) OR ` +
          `cols(${dataColumns} != ${dataColumnsCheck})`
      )
    }

    const rows = headerRows + dataRows
    const columns = headerColumns + dataColumns

    return {
      headerRows,
      headerColumns,
      dataRows,
      dataColumns,
      rows,
      columns,
    }
  }

  public getCell(rowIndex: number, columnIndex: number): TableCell {
    const { headerRows, headerColumns, rows, columns } = this.dimensions

    if (rowIndex < 0 || rowIndex >= rows) {
      throw new Error("Row index is out of range.")
    }
    if (columnIndex < 0 || columnIndex >= columns) {
      throw new Error("Column index is out of range.")
    }

    const isBlankCell = rowIndex < headerRows && columnIndex < headerColumns
    const isIndexCell = rowIndex >= headerRows && columnIndex < headerColumns
    const isColumnsCell = rowIndex < headerRows && columnIndex >= headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push(`level${rowIndex}`)
      }
      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    }

    if (isIndexCell) {
      const dataRowIndex = rowIndex - headerRows

      const uuid = this.styler?.uuid
      const classNames = [
        `row_heading`,
        `level${columnIndex}`,
        `row${dataRowIndex}`,
      ]

      return {
        type: "index",
        id: uuid
          ? `T_${uuid}level${columnIndex}_row${dataRowIndex}`
          : undefined,
        classNames: classNames.join(" "),
        // Table index is stored as is (in the column format).
        content: this.index[columnIndex][dataRowIndex],
      }
    }

    if (isColumnsCell) {
      const dataColumnIndex = columnIndex - headerColumns

      const classNames = [
        `col_heading`,
        `level${rowIndex}`,
        `col${dataColumnIndex}`,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.columns[rowIndex][dataColumnIndex],
      }
    }

    const dataRowIndex = rowIndex - headerRows
    const dataColumnIndex = columnIndex - headerColumns

    const uuid = this.styler?.uuid
    const classNames = ["data", `row${dataRowIndex}`, `col${dataColumnIndex}`]

    const content = this.styler?.displayValues
      ? this.styler.displayValues.getCell(rowIndex, columnIndex).content
      : this.data[dataRowIndex][dataColumnIndex]

    return {
      type: "data",
      id: uuid
        ? `T_${uuid}row${dataRowIndex}_col${dataColumnIndex}`
        : undefined,
      classNames: classNames.join(" "),
      content,
    }
  }
}
