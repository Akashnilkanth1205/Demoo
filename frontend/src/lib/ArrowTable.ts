/**
 * @license
 * Copyright 2018-2019 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Table } from "apache-arrow"
import { Map as ImmutableMap } from "immutable"

type CellType = "blank" | "index" | "columns" | "data"

interface Cell {
  classNames: string
  content: string
  id?: string
  type: CellType
}

export default class ArrowTable {
  private readonly dataTable: Table
  private readonly indexTable: Table
  private readonly columnsTable: Table
  private readonly styler?: ImmutableMap<string, any>

  constructor(
    dataBuffer: Uint8Array,
    indexBuffer: Uint8Array,
    columnsBuffer: Uint8Array,
    styler?: any
  ) {
    this.dataTable = Table.from(dataBuffer)
    this.indexTable = Table.from(indexBuffer)
    this.columnsTable = Table.from(columnsBuffer)
    this.styler = styler
  }

  get rows(): number {
    console.log(
      "rows",
      this.indexTable.length,
      this.columnsTable.numCols,
      this.columnsTable.length,
      this.columnsTable.schema.fields
    )
    return this.indexTable.length + this.columnsTable.numCols
  }

  get columns(): number {
    return this.indexTable.numCols + this.columnsTable.length
  }

  get headerRows(): number {
    return this.rows - this.dataRows
  }

  get headerColumns(): number {
    return this.columns - this.dataColumns
  }

  get dataRows(): number {
    return this.dataTable.length
  }

  get dataColumns(): number {
    return this.dataTable.numCols
  }

  get uuid(): string | null {
    return this.styler?.get("uuid", null)
  }

  get caption(): string | null {
    return this.styler?.get("caption", null)
  }

  get styles(): string | null {
    return this.styler?.get("styles", null)
  }

  public getCell = (rowIndex: number, columnIndex: number): Cell => {
    const isBlankCell =
      rowIndex < this.headerRows && columnIndex < this.headerColumns
    const isIndexCell =
      rowIndex >= this.headerRows && columnIndex < this.headerColumns
    const isColumnsCell =
      rowIndex < this.headerRows && columnIndex >= this.headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push("level" + rowIndex)
      }

      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    } else if (isColumnsCell) {
      const dataColumnIndex = columnIndex - this.headerColumns
      const classNames = [
        "col_heading",
        "level" + rowIndex,
        "col" + dataColumnIndex,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.getContent(this.columnsTable, dataColumnIndex, rowIndex),
      }
    } else if (isIndexCell) {
      const dataRowIndex = rowIndex - this.headerRows
      const classNames = [
        "row_heading",
        "level" + columnIndex,
        "row" + dataRowIndex,
      ]

      return {
        type: "index",
        id: `T_${this.uuid}level${columnIndex}_row${dataRowIndex}`,
        classNames: classNames.join(" "),
        content: this.getContent(this.indexTable, dataRowIndex, columnIndex),
      }
    } else {
      const dataRowIndex = rowIndex - this.headerRows
      const dataColumnIndex = columnIndex - this.headerColumns
      const classNames = ["data", "row" + dataRowIndex, "col" + dataColumnIndex]

      return {
        type: "data",
        id: `T_${this.uuid}row${dataRowIndex}_col${dataColumnIndex}`,
        classNames: classNames.join(" "),
        content: this.getContent(this.dataTable, dataRowIndex, dataColumnIndex),
      }
    }
  }

  private getContent = (
    table: Table,
    rowIndex: number,
    columnIndex: number
  ): string => {
    const column = table.getColumnAt(columnIndex)
    if (column === null) {
      return ""
    }
    return column.get(rowIndex)
  }
}
