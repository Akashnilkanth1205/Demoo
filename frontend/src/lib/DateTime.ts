/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import moment from "moment-timezone"
import { Datetime } from "autogen/proto"

class DateTimeHandler {
  static moment(date: any, timezone?: string): moment.Moment {
    const nanos = date.datetime || (date.get && date.get("datetime"))
    let datetime = moment(nanos ? this.nanosToMilli(nanos) : date)
    const datetimeTimezone =
      timezone || date.timezone || (date.get && date.get("timezone"))
    return datetimeTimezone
      ? datetime.tz(datetimeTimezone)
      : datetime.tz("UTC").tz(moment.tz.guess(), true)
  }

  static nanosToMilli(nanos: number): number {
    return nanos / 1e6
  }

  static nanosToDuration(nanos: number): Duration {
    return new Duration(nanos / 1e6)
  }

  static dateToString(date: any): string {
    date = this.moment(date)
    let format = "lll z"
    if (date.hour() === 0 && date.minute() === 0 && date.second() === 0) {
      format = "ll z"
    }

    return date.format(format)
  }
}

class Duration {
  private readonly millis: number

  constructor(millis: number) {
    this.millis = millis
  }

  getTime(): number {
    return this.millis
  }
}

export { DateTimeHandler, Duration }
