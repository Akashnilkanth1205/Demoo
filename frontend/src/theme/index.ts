/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LightTheme, DarkTheme } from "baseui"
import { lightBaseUITheme, darkBaseUITheme } from "./baseui"
import base from "./baseTheme"
import light from "./lightTheme"
import dark from "./darkTheme"
import sidebar from "./sidebarTheme"
import { Theme } from "./types"

export * from "./baseui"
export * from "./types"
export * from "./utils"
export const lightTheme: Theme = light
export const darkTheme: Theme = dark
export const sidebarTheme: Theme = sidebar

export const AvailableTheme = {
  lightTheme: {
    name: "Light",
    emotion: light,
    baseweb: LightTheme,
    basewebTheme: lightBaseUITheme,
  },
  darkTheme: {
    name: "Dark",
    emotion: dark,
    baseweb: DarkTheme,
    basewebTheme: darkBaseUITheme,
  },
  customTheme: {
    // TODO: base, main or nothing?
    name: "Custom",
    emotion: base,
    baseweb: LightTheme,
    basewebTheme: lightBaseUITheme,
  },
}
