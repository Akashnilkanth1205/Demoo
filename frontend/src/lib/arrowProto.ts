/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"
import { Data, Index, IndexType, Quiver } from "src/lib/Quiver"

export interface DataFrame {
  index: Index
  columns: string[][]
  data: Data
  styler?: Styler
}

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface IndexData {
  data: string[]
  type: IndexType
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

export enum IndexTypes {
  UnicodeIndex = "unicode",
  RangeIndex = "range",
  CategoricalIndex = "categorical",
  MultiIndex = "multiIndex",
  IntervalIndex = "interval[int64]",
  DatetimeIndex = "datetime",
  TimedeltaIndex = "time",
  PeriodIndex = "period[Q-DEC]",
  Int64Index = "int64",
  UInt64Index = "uint64",
  Float64Index = "float64",
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: any
}

export function parseTable(table: Table): DataFrame {
  const schema = getSchema(table)
  const index = {
    data: getIndex(table, schema),
    type: getIndexType(schema),
  }
  const columns = getColumns(schema)
  const data = {
    data: getData(table, columns),
    type: getColumnType(schema),
  }
  return {
    index,
    columns,
    data,
  }
}

function getSchema(table: Table): Schema {
  const schema = table.schema.metadata.get("pandas")
  if (schema == null) {
    throw new Error("Table schema is missing.")
  }
  return JSON.parse(schema)
}

function getIndex(table: Table, schema: Schema): string[][] {
  return schema.index_columns.map(field => {
    const isRangeIndex = typeof field === "object" && field.kind === "range"
    if (isRangeIndex) {
      const { start, stop, step } = field as RangeIndex
      return range(start, stop, step).map(String)
    }
    const column = table.getColumn(field as string)
    return getColumnData(column).map(String)
  })
}

function getIndexType(schema: Schema): IndexType[] {
  return schema.index_columns.map(indexName => {
    if (isRangeIndex(indexName)) {
      return {
        name: IndexTypes.RangeIndex,
        meta: indexName,
      }
    }

    // Get index column from columns schema.
    const indexColumn = schema.columns.find(
      column => column.field_name === indexName
    )

    // PeriodIndex and IntervalIndex values are kept in `numpy_type` property,
    // the rest in `pandas_type`.
    return {
      name:
        indexColumn?.pandas_type === "object"
          ? indexColumn?.numpy_type
          : indexColumn?.pandas_type,
      meta: indexColumn?.metadata,
    }
  })
}

function getColumns(schema: Schema): string[][] {
  const isMultiIndex = schema.column_indexes.length > 1
  return unzip(
    schema.columns
      .map(column => column.field_name)
      .filter(fieldName => !schema.index_columns.includes(fieldName))
      .map(fieldName =>
        isMultiIndex
          ? JSON.parse(
              fieldName
                .replace(/\(/g, "[")
                .replace(/\)/g, "]")
                .replace(/'/g, '"')
            )
          : [fieldName]
      )
  )
}

function getColumnType(schema: Schema): any {
  return schema.columns
    .filter(column => !schema.index_columns.includes(column.field_name))
    .map(column => column.pandas_type)
}

function getData(table: Table, columns: string[][]): string[][] {
  const rows = table.length
  const cols = columns.length > 0 ? columns[0].length : 0

  return range(0, rows).map(rowIndex =>
    range(0, cols).map(columnIndex =>
      table.getColumnAt(columnIndex)?.get(rowIndex)
    )
  )
}

function getColumnData(column: Column): string[] {
  return range(0, column.length).map(rowIndex => column.get(rowIndex))
}

function isRangeIndex(field: string | RangeIndex): boolean {
  return typeof field === "object" && field.kind === "range"
}

export function concatTables(q1: Quiver, q2: Quiver): DataFrame {
  // Special case if df1 is empty.
  if (q1.data.data.length === 0) {
    return {
      index: q2.index,
      data: q2.data,
      columns: q2.columns,
    }
  }

  // Always return df1 columns
  const { columns } = q1
  const index = concatIndices(q1.index, q2.index)
  const data = concatData(q1.data, q2.data)
  return { index, data, columns }
}

function concatIndices(i1: Index, i2: Index): Index {
  // Special case if i1 is empty.
  if (i1.data.length === 0) {
    return i2
  }

  // Otherwise, make sure the types match.
  if (!sameIndexTypes(i1.type, i2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(i1.type)} with ${JSON.stringify(
        i2.type
      )}.`
    )
  }

  return i1.data.reduce(
    (newIndex: Index, indexData: string[], index: number) => {
      const concatenatedIndex = concatIndex(
        indexData,
        i2.data[index],
        i1.type[index]
      )
      newIndex.data.push(concatenatedIndex.data)
      newIndex.type.push(concatenatedIndex.type)
      return newIndex
    },
    { data: [], type: [] }
  )
}

function concatIndex(i1: string[], i2: string[], type: IndexType): IndexData {
  // Special case for RangeIndex.
  if (type.name === IndexTypes.RangeIndex) {
    return concatRangeIndex(i1, i2, type)
  }
  return concatAnyIndex(i1, i2, type)
}

function concatRangeIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  let newStop = type.meta.stop
  return i2.reduce(
    (newIndex: IndexData) => {
      newIndex.data.push(newStop)
      newStop += type.meta.step
      newIndex.type.meta.stop = newStop
      return newIndex
    },
    { data: i1, type }
  )
}

function concatAnyIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  const concatenatedIndex = i1.concat(i2)

  // Special case for CategoricalIndex:
  // `num_categories` must be increased.
  if (type.name === IndexTypes.CategoricalIndex) {
    type.meta.num_categories += i2.length
  }

  return { data: concatenatedIndex, type }
}

function concatData(d1: Data, d2: Data): Data {
  // Make sure data types match.
  if (!sameColumnTypes(d1.type, d2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(d1.type)} with ${JSON.stringify(
        d2.type
      )}.`
    )
  }

  const numberOfColumns = d1.type.length
  return {
    data: d1.data.concat(
      d2.data.map((data: string[]) => data.slice(0, numberOfColumns))
    ),
    type: d1.type,
  }
}

function sameIndexTypes(t1: IndexType[], t2: IndexType[]): boolean {
  // Make sure both indices have same dimensions.
  if (t1.length !== t2.length) {
    return false
  }

  return t1.every(
    (firstTypeValue: IndexType, index: number) =>
      firstTypeValue.name === t2[index].name
  )
}

function sameColumnTypes(firstType: string[], secondType: string[]): boolean {
  return firstType.every(
    (firstTypeValue: string, index: number) =>
      firstTypeValue === secondType[index]
  )
}
