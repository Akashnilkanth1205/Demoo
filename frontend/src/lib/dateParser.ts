import parser, { Parser, Eat, Tokenizer } from "remark-parse"
import { DateTimeHandler } from "lib/DateTime"
import { Datetime } from "autogen/proto"
function dates() {
  //@ts-ignore
  var Parser = this.Parser
  var tokenizers = Parser.prototype.inlineTokenizers
  var methods = Parser.prototype.inlineMethods

  tokenizers.datetime = tokenizeDatetimes as Tokenizer
  // Run it just before `text`.
  methods.splice(methods.indexOf("text"), 0, "datetime")
}

tokenizeDatetimes.notInLink = true
tokenizeDatetimes.locator = locateDate

function locateDate(value: string, fromIndex: number) {
  return value.indexOf("{@", fromIndex)
}

function tokenizeDatetimes(eat: Eat, value: string, silent: boolean) {
  var match = /{@datetime: (.*)}(.*$)/.exec(value)
  if (match) {
    if (silent) {
      return true
    }

    const date = Datetime.fromObject(JSON.parse(match[1]))
    console.log(JSON.parse(match[1]), date)
    // Reconstruct markdown
    const markdownParts = [
      { type: "inlineCode", value: DateTimeHandler.dateToString(date) },
    ]
    if (match.index) {
      markdownParts.unshift({
        type: "text",
        value: value.substring(0, match.index),
      })
    }
    if (match[2]) {
      markdownParts.push({ type: "text", value: match[2] })
    }

    return eat(value)({
      type: "root",
      children: markdownParts,
    })
  }
}

export default dates
