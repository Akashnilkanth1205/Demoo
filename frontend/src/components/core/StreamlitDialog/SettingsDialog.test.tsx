/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { mainTheme, darkTheme } from "theme"
import { mount, shallow } from "lib/test_util"

import { SettingsDialog, Props } from "./SettingsDialog"

const getProps = (extend?: Partial<Props>): Props => ({
  isServerConnected: true,
  onClose: jest.fn(),
  onSave: jest.fn(),
  settings: { wideMode: false, runOnSave: false, activeTheme: mainTheme },
  allowRunOnSave: false,
  allowedThemes: [],
  ...extend,
})

describe("SettingsDialog", () => {
  it("renders without crashing", () => {
    const props = getProps()
    const wrapper = shallow(<SettingsDialog {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it("should render run on save", () => {
    const props = getProps({
      allowRunOnSave: true,
    })

    const wrapper = mount(<SettingsDialog {...props} />)
    const checkboxes = wrapper.find("input[type='checkbox']")

    expect(checkboxes).toHaveLength(2)
    expect(wrapper.state("runOnSave")).toBe(false)

    checkboxes
      .at(0)
      .simulate("change", { target: { name: "runOnSave", checked: true } })
    wrapper.update()

    expect(wrapper.state("runOnSave")).toBe(true)
  })

  it("should render allowedThemes", () => {
    const allowedThemes = [mainTheme, darkTheme]
    const props = getProps({ allowedThemes })
    const wrapper = mount(<SettingsDialog {...props} />)
    const radioBtns = wrapper.find("Radio").slice(1)

    expect(radioBtns).toHaveLength(2)

    const radioLabels = radioBtns.map(btn => btn.prop("children"))
    expect(radioLabels).toEqual(allowedThemes.map(theme => theme.name))

    wrapper
      .find("input[type='radio']")
      .at(1)
      .simulate("change", { target: { value: "1" } })
    wrapper.update()
    expect(wrapper.state("activeTheme")).toEqual(darkTheme)
  })
})
