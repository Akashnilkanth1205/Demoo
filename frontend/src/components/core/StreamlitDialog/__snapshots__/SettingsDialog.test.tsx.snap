// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsDialog renders without crashing 1`] = `
<Modal
  isOpen={true}
  onClose={[Function]}
>
  <ModalHeader>
    Settings
  </ModalHeader>
  <ModalBody>
    <h3>
      Appearance
    </h3>
    <label>
      <input
        checked={false}
        name="wideMode"
        onChange={[Function]}
        type="checkbox"
      />
       
      Wide mode
    </label>
    <div>
      <StyledSmall>
        Turn on to make this app occupy the entire width of the screen
      </StyledSmall>
    </div>
    <StyledLabel>
      Theme
    </StyledLabel>
    <StyledSmall>
      Choose app and font colors/styles
    </StyledSmall>
    <Selectbox
      disabled={false}
      onChange={[Function]}
      options={
        Array [
          "Light",
          "Dark",
        ]
      }
      value={0}
    />
    <ThemeCreator
      label="Edit Existing Custom Theme"
      themeInput={
        Object {
          "backgroundColor": "#FFFFFF",
          "font": "IBM Plex Sans, sans-serif",
          "name": "Light",
          "primaryColor": "#f63366",
          "secondaryBackgroundColor": "#f0f2f6",
          "secondaryColor": "#a3a8b8",
          "textColor": "#262730",
        }
      }
      updateThemeInput={[Function]}
    />
  </ModalBody>
</Modal>
`;
